pre_commit/languages/rust.py

```diff
@@ -16,24 +22,56 @@
 from pre_commit.prefix import Prefix
 from pre_commit.util import clean_path_on_failure
 from pre_commit.util import cmd_output_b
+from pre_commit.util import make_executable
+from pre_commit.util import win_exe
 
 ENVIRONMENT_DIR = 'rustenv'
-get_default_version = helpers.basic_get_default_version
+TOOLCHAIN_FROM_VERSION = {
+    'system': None,
+    C.DEFAULT: 'stable',
+}
+
 health_check = helpers.basic_health_check
 
 
-def get_env_patch(target_dir: str) -> PatchesT:
+@functools.lru_cache(maxsize=1)
+def get_default_version() -> str:
+    # If rust is already installed, we can save a bunch of setup time by
+    # using the installed version.
+    #
+    # Just detecting the executable does not suffice, because if rustup is
+    # installed but no toolchain is available, then `cargo` exists but
+    # cannot be used without installing a toolchain first.
+    retcode, _, _ = cmd_output_b('cargo', '--version', retcode=0)
+    if retcode == 0:
+        return 'system'
+    return C.DEFAULT
+
+
+def _envdir(prefix: Prefix, version: str) -> str:
+    directory = helpers.environment_dir(ENVIRONMENT_DIR, version)
+    return prefix.path(directory)
+
+
+def get_env_patch(target_dir: str, version: str) -> PatchesT:
+    toolchain = TOOLCHAIN_FROM_VERSION.get(version, version)
     return (
+        ('CARGO_HOME', target_dir),
         ('PATH', (os.path.join(target_dir, 'bin'), os.pathsep, Var('PATH'))),
+        # Only set RUSTUP_TOOLCHAIN if we don't want use the system's default
+        # toolchain
+        *((('RUSTUP_TOOLCHAIN', toolchain),) if toolchain else ()),
```

Comment:
this then becomes `if version != 'system' else ()`

Reply:
https://github.com/pre-commit/pre-commit/commit/[8d545ad](https://github.com/pre-commit/pre-commit/commit/8d545adedbcf33c731a995f7b2fe22fa1be2bef4)edbcf33c731a995f7b2fe22fa1be2bef4